storage/innobase/buf/buf0flu.cc

```diff
@@ -2024,14 +2025,20 @@ ATTRIBUTE_COLD void buf_flush_ahead(lsn_t lsn, bool furious)
 
 /** Conduct checkpoint-related flushing for innodb_flush_sync=ON,
 and try to initiate checkpoints until the target is met.
-@param lsn   minimum value of buf_pool.get_oldest_modification(LSN_MAX) */
-ATTRIBUTE_COLD static void buf_flush_sync_for_checkpoint(lsn_t lsn)
+@param lsn   minimum value of buf_pool.get_oldest_modification(LSN_MAX)
+@param cbk   callback to caller before a flush iteration. */
+ATTRIBUTE_COLD
+static void buf_flush_sync_for_checkpoint(lsn_t lsn, std::function<void()> cbk)
 {
   ut_ad(!srv_read_only_mode);
   mysql_mutex_assert_not_owner(&buf_pool.flush_list_mutex);
 
   for (;;)
   {
+    /* Callback before every iteration. Currently used to check and free LRU
+    pages when buffer pool is running out of free pages. */
+    cbk();
```

Comment:
There is only one caller of `buf_flush_sync_for_checkpoint()`, and it is always passing the same function `lru_check` as the callback. That function is not being invoked anywhere else.

To avoid any runtime overhead, I would suggest to define `lru_check` as a `static` function right before this function, so that the compiler will be able to inline the invocation. Its name could be `buf_flush_sync_LRU` or something similar.

I don’t see any way to reuse that function in `buf_flush_page_cleaner()`. The conditions for invoking `buf_flush_LRU()` are slightly different in the case of non-furious flushing. The only thing we might do is to move the mutex release inside `buf_flush_LRU()` and maybe rename that function to `buf_flush_and_release_LRU()` or similar to indicate that the `buf_pool.mutex` will be released.

Reply:
Isn't the dynamic overhead should be negligible here ? The reason for having the callback is to abstract the LRU flush check etc. from the rest of the flow. If you do think dynamic overhead is more, please re-consider my previous implementation.

Reply:
I am all for abstractions that make code more readable, preferrably zero overhead abstractions. In this case, there seems to be nontrivial overhead (beyond what I would have expected):
```asm
   0x0000000001178b77 <+23>:	cmpq   $0x0,0x10(%r12)
   0x0000000001178b7d <+29>:	jne    0x1178b84 <buf_flush_sync_for_checkpoint(lsn_t, std::function<void()>)+36>
   0x0000000001178b7f <+31>:	call   0x13f6a60 <_ZSt25__throw_bad_function_callv@plt>
   0x0000000001178b84 <+36>:	mov    %r12,%rdi
   0x0000000001178b87 <+39>:	call   *0x18(%r12)
```
In addition to this we would of course have some function preamble and cleanup overhead. The parameter `cbk` also needs to be preserved in the loop.

Because of this, I checked the previous approach 0c209095660542f20a84a9ccceaeed880f58e86c again. I like it better. Basically, it removes the `for` loop from `buf_flush_sync_for_checkpoint()` and adds some code to its only caller.

I would suggest the moving back to more like 0c209095660542f20a84a9ccceaeed880f58e86c:
1. Add the needed code to the start of `buf_flush_sync_for_checkpoint()` and do not add anything else than some debug injection to `buf_flush_page_cleaner()`.
1. Consider declaring the function `buf_flush_sync_for_checkpoint()` not only `ATTRIBUTE_COLD` but also `ATTRIBUTE_NOINLINE`. I do not know why and I did not test other compilers, but `g++-14` would inline the cold function invocation, which was not my intention.

Reply:
ok, modified the way you suggested.